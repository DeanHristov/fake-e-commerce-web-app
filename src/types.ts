export enum PRODUCT_CATEGORIES {
  ELECTRONIC = 'ELECTRONIC',
}

export type TCurrency = 'USD' | 'BGN' | 'EUR';

export interface IComment {
  productId: string;
  _id: string;
  name: string;
  email: string;
  body: string;
  date: string;
}

// Object schema can be seen here: https://dummyjson.com/docs/products
export interface IProduct {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images?: string[];
  comments?: IComment[];
}

export interface IUser {
  name: string;
  email: string;
  password: string;
  isAdmin: boolean;
}

export interface ICartProduct {
  id: number;
  price: number;
  quantity: number;
  discountPercentage?: number;
  // TODO Fix next line!
  discountedPrice?: number;
  title: string;
  thumbnail: string;
  total: number;
  inStock: number;
}

export interface ICart {
  id: number;
  userId: number;
  total: number;
  discountedTotal: number;
  totalProducts: number;
  totalQuantity: number;
  currency?: TCurrency;
  products: Record<number, ICartProduct>; // Map<number, ICartProduct>;
}

//@see:https://nextjs.org/docs/app/api-reference/functions/generate-metadata#generatemetadata-function
export interface IMetadataPageTitle {
  title: string;
}

export interface IProductsResponse {
  products: IProduct[];
  total: number;
  skip: number;
  limit: number;
}

export interface ICartsResponse {
  carts: ICart[];
  total: number;
  skip: number;
  limit: number;
}
