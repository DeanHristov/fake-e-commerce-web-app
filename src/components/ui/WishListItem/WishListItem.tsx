import Button from '@/components/ui/Button';
import PriceLabel from '@/components/ui/PriceLabel';
import ProductHeading from '@/components/ui/ProductHeading';
import StarRating from '@/components/ui/StarRating';
import { useDispatch } from '@/store';
import { addToCart, removeFromWishList } from '@/store/slices';
import { ICartProduct } from '@/types';
import { ShoppingCartIcon, TrashIcon } from '@heroicons/react/24/solid';
import Image from 'next/image';
import { FC, useMemo } from 'react';

export interface IWishListItemProps {
  product: ICartProduct;
}

const WishListItem: FC<IWishListItemProps> = ({ product }) => {
  const dispatch = useDispatch();
  const productItem: ICartProduct = useMemo(
    () => ({
      _id: product._id,
      title: product.title,
      price: product.price,
      total: product.price,
      inStock: product.inStock,
      quantity: 1,
      discountPercentage: product.discountPercentage,
      discountedPrice: product.price,
      thumbnail: product.thumbnail,
    }),
    [product],
  );

  return (
    <div className="relative border border-gray-300 px-2 py-2 my-2 flex justify-center items-center bg-white rounded-lg shadow-md gap-x-2 hover:scale-[1.02] transition">
      <Image
        priority
        src={product.thumbnail}
        alt={product.title}
        width={0}
        height={0}
        sizes="100vw"
        className={`mx-auto h-auto w-[8rem] max-h-[5rem] rounded-lg transition border border-gray-300`}
      />
      <div className="flex-1 justify-center items-center">
        <ProductHeading
          id={product._id}
          title={product.title}
          className="inline-block"
        />
        <StarRating
          activeColor="#FFA133"
          inActiveColor="#E3E3E3"
          starCount={5}
          rating={4}
        />
      </div>
      <div className="absolute top-1/2 -translate-y-1/2 right-36">
        <PriceLabel
          currency={'EUR'}
          discountPercentage={product.discountPercentage}
          price={product.price}
        />
      </div>
      <div className="w-32 space-y-2">
        <Button
          variant="secondary"
          onClick={() => dispatch(addToCart(productItem))}
          leftIcon={<ShoppingCartIcon className={'w-4 h-4 text-dark'} />}
        />

        <Button
          variant="secondary"
          onClick={() =>
            dispatch(removeFromWishList({ productId: product._id }))
          }
          leftIcon={<TrashIcon className={'w-4 h-4 text-dark'} />}
        />
      </div>
    </div>
  );
};

export default WishListItem;
