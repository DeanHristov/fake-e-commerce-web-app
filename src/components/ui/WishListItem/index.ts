import WishListItem, { IWishListViewProps } from './WishListItem';

export type { IWishListViewProps };
export default WishListItem;
