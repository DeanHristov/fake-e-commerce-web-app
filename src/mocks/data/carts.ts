import { ICart } from '@/types';

export const carts: ICart[] = [
  {
    _id: '34e1d41b-384a-4bf4-866e-ab5b268d124c',
    products: {
      '8f12eb53-18c8-46a6-b975-ea0ee8b62d4c': {
        _id: '8f12eb53-18c8-46a6-b975-ea0ee8b62d4c',
        title: 'Spring and summershoes',
        price: 20,
        quantity: 3,
        total: 60,
        discountPercentage: 8.71,
        discountedPrice: 55,
        thumbnail: 'https://i.dummyjson.com/data/products/59/thumbnail.jpg',
        inStock: 5,
      },
    },
    currency: 'EUR',
    total: 2,
    discountedTotal: 1941,
    userId: 'ba500896-80ca-42bd-b7f7-a743b959203e',
    totalProducts: 2,
    totalQuantity: 2,
  },
];
