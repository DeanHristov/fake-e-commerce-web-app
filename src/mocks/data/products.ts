import { IProduct, PRODUCT_CATEGORIES } from '../../types';

const products: IProduct[] = [
  {
    _id: '29a34b48-8343-4e3d-8c21-3548ba249936',
    name: 'Airpods Pro2 Headphones',
    image: '/images/airpods.jpg',
    description:
      "Apple's new H2 chip is the driving force behind AirPods Pro and their improved sound characteristics. It works in unison with a personalized driver and amplifier for crystal clear heights and deep, rich bass with incredible clarity, so every sound is more pronounced than ever",
    brand: 'Apple',
    category: PRODUCT_CATEGORIES.ELECTRONIC,
    price: 399.99,
    countInStock: 10,
    rating: 4.5,
    currency: 'USD',
    reviews: 5,
  },
];

export default products;
