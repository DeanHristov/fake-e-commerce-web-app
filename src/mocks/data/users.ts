import bcrypt from 'bcryptjs';
import { IUser } from '../../types';

const users: IUser[] = [
  {
    name: 'Admin User',
    email: 'admin@example.com',
    password: bcrypt.hashSync('123456', 10),
    isAdmin: true,
  },
  {
    name: 'Regular User',
    email: 'regular-user@example.com',
    password: bcrypt.hashSync('123456', 10),
    isAdmin: false,
  },
];

export default users;
