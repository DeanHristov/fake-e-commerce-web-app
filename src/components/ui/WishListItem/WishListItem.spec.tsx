import WishListItem from '@/components/ui/WishListItem/index';
import { carts } from '@/mocks/data/carts';
import { ICartProduct } from '@/types';
import { render } from '@testing-library/react';

const firstProductItem: ICartProduct =
  carts[0].products['8f12eb53-18c8-46a6-b975-ea0ee8b62d4c'];

describe('UI/Component <WishListItem {...} />', () => {
  it('Should render the default state', () => {
    const { container } = render(<WishListItem product={firstProductItem} />);

    expect(container).toMatchSnapshot();
  });
});
