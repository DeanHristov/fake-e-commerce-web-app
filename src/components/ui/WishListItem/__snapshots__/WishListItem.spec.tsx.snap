// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Component <WishListItem {...} /> Should render the default state 1`] = `
<div>
  <div
    class="relative border border-gray-300 px-2 py-2 my-2 flex justify-center items-center bg-white rounded-lg shadow-md gap-x-2 hover:scale-[1.02] transition"
  >
    <img
      alt="Spring and summershoes"
      class="mx-auto h-auto w-[8rem] max-h-[5rem] rounded-lg transition border border-gray-300"
      data-nimg="1"
      decoding="async"
      fetchpriority="high"
      height="0"
      sizes="100vw"
      src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=3840&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=3840&q=75 3840w"
      style="color: transparent;"
      width="0"
    />
    <div
      class="flex-1 justify-center items-center"
    >
      <h3
        class="product-heading hover:underline hover:cursor-pointer inline-block"
      >
        Spring and summershoes
      </h3>
      <div
        class="flex space-x-2 items-center mt-3"
        data-testid="star-rating"
      >
        <div
          class="flex space-x-1 justify-start items-baseline"
        >
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(255, 161, 51);"
          >
            ☆
          </span>
          <span
            class="star-icon"
            data-testid="star-item"
            style="color: rgb(227, 227, 227);"
          >
            ☆
          </span>
        </div>
        <span
          class="text-xs font-medium text-zambezi leading-4"
          data-testid="rating"
        >
          4 Reviews
        </span>
      </div>
    </div>
    <div
      class="absolute top-1/2 -translate-y-1/2 right-36"
    >
      <div
        class="space-x-2"
      >
        <span
          class="price-label-old line-through text-sm"
        >
          €22.00
        </span>
        <span
          class="price-label-old"
        >
          /
        </span>
        <span
          class="price-label"
        >
          €20.00
        </span>
      </div>
    </div>
    <div
      class="w-32 space-y-2"
    >
      <button
        class="cursor-pointer bg-white border border-green-600 text-dark font-normal hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 undefined"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4 text-dark"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2.25 2.25a.75.75 0 000 1.5h1.386c.17 0 .318.114.362.278l2.558 9.592a3.752 3.752 0 00-2.806 3.63c0 .414.336.75.75.75h15.75a.75.75 0 000-1.5H5.378A2.25 2.25 0 017.5 15h11.218a.75.75 0 00.674-.421 60.358 60.358 0 002.96-7.228.75.75 0 00-.525-.965A60.864 60.864 0 005.68 4.509l-.232-.867A1.875 1.875 0 003.636 2.25H2.25zM3.75 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM16.5 20.25a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
          />
        </svg>
      </button>
      <button
        class="cursor-pointer bg-white border border-green-600 text-dark font-normal hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 undefined"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4 text-dark"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16.5 4.478v.227a48.816 48.816 0 013.878.512.75.75 0 11-.256 1.478l-.209-.035-1.005 13.07a3 3 0 01-2.991 2.77H8.084a3 3 0 01-2.991-2.77L4.087 6.66l-.209.035a.75.75 0 01-.256-1.478A48.567 48.567 0 017.5 4.705v-.227c0-1.564 1.213-2.9 2.816-2.951a52.662 52.662 0 013.369 0c1.603.051 2.815 1.387 2.815 2.951zm-6.136-1.452a51.196 51.196 0 013.273 0C14.39 3.05 15 3.684 15 4.478v.113a49.488 49.488 0 00-6 0v-.113c0-.794.609-1.428 1.364-1.452zm-.355 5.945a.75.75 0 10-1.5.058l.347 9a.75.75 0 101.499-.058l-.346-9zm5.48.058a.75.75 0 10-1.498-.058l-.347 9a.75.75 0 001.5.058l.345-9z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
